common/flatpak-transaction.c

```diff
@@ -3024,6 +3031,125 @@ flatpak_transaction_add_install_flatpakref (FlatpakTransaction *self,
+  if (remotes == NULL)
+    return FALSE;
+
+  preinstalled_refs = flatpak_dir_get_config_patterns (priv->dir, "preinstalled");
+
+  preinstall_paths = flatpak_get_preinstall_config_file_paths (NULL, error);
+  if (preinstall_paths == NULL)
+    return FALSE;
+
+  /* Find preinstalls that should get installed */
+  for (int i = 0; preinstall_paths[i] != NULL; i++)
+    {
+      const char *path = preinstall_paths[i];
+      g_autoptr(FlatpakDecomposed) decomposed = NULL;
+      g_autofree char *collection_id = NULL;
+
+      if (!flatpak_parse_preinstall_config_file (path, priv->default_arch, &decomposed, &collection_id, error))
+        return FALSE;
+
+      g_info ("Found preinstall ref %s from config file %s", flatpak_decomposed_get_ref (decomposed), path);
+
+      /* Store for later */
+      g_ptr_array_add (install_refs, flatpak_decomposed_dup_ref (decomposed));
+
+      /* Skip over if it's listed as previously preinstalled - it's now under
+       * user's control and we no longer install it again, even if the user
+       * manually removes it. */
+      if (!priv->reinstall &&
+          flatpak_g_ptr_array_contains_string (preinstalled_refs, flatpak_decomposed_get_ref (decomposed)))
+        {
+          g_info ("Preinstall ref %s is marked as already preinstalled; skipping", flatpak_decomposed_get_ref (decomposed));
+          continue;
+        }
+
+      for (int j = 0; remotes[j] != NULL; j++)
+        {
+          const char *remote = remotes[j];
+          g_autoptr(GError) local_error = NULL;
+          g_autofree char *remote_collection_id = NULL;
+
+          if (flatpak_dir_get_remote_disabled (priv->dir, remote))
+            continue;
+
+          remote_collection_id = flatpak_dir_get_remote_collection_id (priv->dir, remote);
+
+          /* Choose the first match if the collection ID was not specified */
+          if (collection_id != NULL &&
+              g_strcmp0 (remote_collection_id, collection_id) != 0)
+            continue;
+
+          g_info ("Adding preinstall of %s from remote %s", flatpak_decomposed_get_ref (decomposed), remote);
+
+          if (!flatpak_transaction_add_ref (self, remote, decomposed, NULL, NULL, NULL,
+                                            FLATPAK_TRANSACTION_OPERATION_INSTALL,
+                                            NULL, NULL, FALSE, TRUE, NULL,
+                                            &local_error))
+            g_info ("Failed to add preinstall ref %s: %s", flatpak_decomposed_get_ref (decomposed),
+                     local_error->message);
+        }
+    }
+
+  /* Find previously preinstalled refs that are no longer in the preinstall
+   * list and should now get uninstalled */
+  for (guint i = 0; i < preinstalled_refs->len; i++)
+    {
+      const char *ref = g_ptr_array_index (preinstalled_refs, i);
+
+      /* No longer in the preinstall.d list, so uninstall */
+      if (!flatpak_g_ptr_array_contains_string (install_refs, ref))
+        {
```

Comment:
```suggestion
      if (flatpak_g_ptr_array_contains_string (install_refs, ref))
        continue;
```
to keep the nesting down

Reply:
Ah, good idea - let me change this, thanks!